haptools/sim_genotype.py

```diff
@@ -38,54 +37,166 @@ def output_vcf(breakpoints, model_file, vcf, sampleinfo, out):
         if pop in pops:
-            sample_pop[sample] = pop
+            pop_sample[pop].append(sample)
 
     # check if pops are there and if they aren't throw an error
-    assert len(pops) == len(list(set(sample_pop.values())))
-
-    # preprocess breakpoints so we can output line by line
-    # TODO
+    assert sorted(pops) == sorted(list(set(pop_sample.keys())))
+
+    # read VCF to get sample names and their order so we can easily call GTs via their index from their sample. 
+    #      IE sample HG00097 is index 1 in a list of samples [HG00096 HG00097 HG00098]
+    # create sample dictionary that holds sample name to the index in the vcf file for quick access 
+    vcf = VCF(vcf_file)
+    sample_dict = {}
+    for ind, sample in enumerate(vcf.samples):
+        sample_dict[sample] = ind
+
+    # TODO read all chroms in vcf
+
+    # TODO read all chroms in breakpoints and ensure that they each have the same ones
+
+    # create index array to store for every sample which haplotype 
+    # block we are currently processing and choose what samples 
+    # will be used for each haplotype block
+    current_bkps = np.zeros(len(breakpoints), dtype=np.int)
+    hapblock_samples = []
+    for haplotype in breakpoints:
+        hap_samples = []
+        for block in haplotype:
+            sample_name = np.random.choice(pop_sample[block.get_pop()])
+            sample_ind = sample_dict[sample_name]
+            hap_samples.append(sample_ind)
+        hapblock_samples.append(hap_samples)
+
+    # output vcf header to new vcf file we create
+    output_samples = [f"Sample_{hap+1}" for hap in range(int(len(hapblock_samples)/2))]
 
     # Process
-    # Choose starting samples (random choice) in VCF from respective population for each haplotype
+    # Choose starting samples (random choice) in VCF from respective population for each haplotype segment
     #     Also precalculate (random choice) all samples that will be switched too once the current local ancestry block ends.
     # Iterate over VCF and output variants to file(in the beginning write out the header as well) until end of haplotype block for a sample (have to iterate over all samples each time to check)
     # Once VCF is complete we've output everything we wanted
+    # VCF output have a FORMAT field where under format is GT:POP and our sample output is GT:POP ie 1|1:YRI|CEU
+    _write_vcf(breakpoints, hapblock_samples, current_bkps, output_samples, vcf, out+".vcf")
+    return
+
+def _write_vcf(breakpoints, hapblock_samples, current_bkps, out_samples, in_vcf, out_vcf):
+    """
+    in_vcf = cyvcf2 variants we are reading in
+    out_vcf = output vcf file we output too
+    """
+    # output vcf file
+    write_vcf = VariantFile(out_vcf, mode="w")
+
+    # make sure the header is properly structured
+    for contig in range(23):
+        write_vcf.header.contigs.add(str(contig+1))
+    for sample in out_samples:
+        write_vcf.header.add_sample(sample)
```

Comment:
this operation is not linear in time, so it will be really slow for large numbers of samples
(See https://github.com/pysam-developers/pysam/issues/1104 for more details)

As of pysam v0.19.1, there is a new `VariantHeader.add_samples()` method that allows you to add multiple samples at once, so you could try using it first:
```suggestion
    try:
        write_vcf.header.add_samples(out_samples)
    except AttributeError:
        for sample in out_samples:
            write_vcf.header.add_sample(sample)
```